Add code to download bogon /drop files

dlangille opened the pull request:
re: https://www.spamhaus.org/blocklists/do-not-route-or-peer/

dlangille committed:
Add code to download bogon /drop files

re: https://www.spamhaus.org/blocklists/do-not-route-or-peer/

dlangille commented:
Also, see how I was able to set YES/NO in periodic.conf using the trenary operator. I figure that would reduce code.

vbotka commented on tasks/drop.lasso.yml:
```diff
@@ -0,0 +1,37 @@
+- name: "drop.lasso: Install periodic script"
+  ansible.builtin.template:
+    dest: /usr/local/etc/periodic/daily/900.drop-lasso
+    src: "{{ pf_drop_lasso_template }}"
```
```suggestion
    src: "{{ pf_drop_lasso_template.J2 }}"
```

vbotka replied:
The extension is missing, I think.

vbotka commented on tasks/drop.lasso.yml:
```diff
@@ -0,0 +1,37 @@
+- name: "drop.lasso: Enable periodic script"
+  sysrc:
+    name:  daily_pf_droplasso_enable
+    state: present
```
```suggestion
```

vbotka replied:
State `present` is the default.

vbotka commented on tasks/drop.lasso.yml:
```diff
@@ -0,0 +1,37 @@
+- name: "drop.lasso: Enable v4"
+  sysrc:
+    name:  daily_pf_droplasso_v4
+    state: present
```
```suggestion
```

vbotka commented on tasks/drop.lasso.yml:
```diff
@@ -0,0 +1,37 @@
+- name: "drop.lasso: Enable v6"
+  sysrc:
+    name:  daily_pf_droplasso_v6
+    state: present
```
```suggestion
```

vbotka commented on tasks/drop.lasso.yml:
```diff
@@ -0,0 +1,37 @@
+    value: "{{ pf_drop_lasso_v6 | bool | ternary('YES', 'NO') }}"
+    path:  /etc/periodic.conf
+  when: pf_drop_lasso_enable | bool
+  tags: periodic
```
```suggestion
```

vbotka replied:
I think this tag can be removed.

vbotka commented on tasks/drop.lasso.yml:
```diff
@@ -0,0 +1,37 @@
+  when: pf_drop_lasso_enable | bool
+  
+- name: "drop.lasso: Enable periodic script"
+  sysrc:
```
```suggestion
  community.general.sysrc:
```

vbotka replied:
Let's use FQCN. Otherwise, *ansible-lint* complains `syntax-check[unknown-module]`
```suggestion
  community.general.sysrc:
```

vbotka commented on tasks/drop.lasso.yml:
```diff
@@ -0,0 +1,37 @@
+    path:  /etc/periodic.conf
+
+- name: "drop.lasso: Enable v4"
+  sysrc:
```
```suggestion
  community.general.sysrc:
```

vbotka commented on tasks/drop.lasso.yml:
```diff
@@ -0,0 +1,37 @@
+  when: pf_drop_lasso_enable | bool
+
+- name: "drop.lasso: Enable v6"
+  sysrc:
```
```suggestion
  community.general.sysrc:
```

vbotka commented on defaults/main/drop_lasso.yml:
```diff
@@ -0,0 +1,13 @@
+# see https://www.spamhaus.org/drop/
+#
+pf_drop_lasso_v4: True
+pf_drop_lasso_v6: True
```
Let's use consistently lowercase boolean across the role.

```suggestion
pf_drop_lasso_v4: true
pf_drop_lasso_v6: true
```

dlangille committed:
Suggested changes

dlangille commented:
I have one more change:

I think the `drop.lasso` periodic values should optionally go to `/etc/periodic.conf.local` (with a knob to decide which location).

Why? Some installations, including mine, distribute the same `/etc/periodic.conf` file to each host. Local stuff goes into `.local`

dlangille committed:
Allow the user to specify the names for the drop.lasso files
The download script contains default values
We aim to not add those filename configuration to perioidic config
unless they are non-default values. (i.e. user specified)

It's not ideal: at present, the default file names are hardcoded in
two places:

  tasks/drop.lasso.yml
  templates/default-drop-lasso.j2

I thought of this alternative:
* define the default filenames
* leave the user specified values undefined
* this means the script has a templated default filename, not hard code
* same for the tasks

I was aiming for default values which the user could not alter, but
could override.

dlangille commented:
Please don't push yet. Still thinking about bogon sources.

vbotka commented:
> I have one more change:
> I think the `drop.lasso` periodic values should optionally go to `/etc/periodic.conf.local` (with a knob to decide which location).

Yes. I think this will be useful.

vbotka commented:
> Please don't push yet. Still thinking about bogon sources.

Sure. I have more suggestions too.

vbotka commented on defaults/main/drop_lasso.yml:
```diff
@@ -0,0 +1,25 @@
+#
+pf_drop_lasso_v4: true
+pf_drop_lasso_v6: true
+
```
The configuration can be simplified further

```suggestion
pf_drop_lasso_rcconf:
  - { name: daily_pf_droplasso_enable, value: "{{ pf_drop_lasso_enable }}" }
  - { name: daily_pf_droplasso_v4, value: "{{ pf_drop_lasso_v4 | bool | ternary('YES', 'NO') }}" }
  - { name: daily_pf_droplasso_v6, value: "{{ pf_drop_lasso_v6 | bool | ternary('YES', 'NO') }}" }
    # these variables are also set in templates/default-drop-lasso.j2
  - { name: daily_pf_droplasso_v4_filename, value: "{{ pf_drop_lasso_v4_filename | default('/etc/pf.drop_v4.txt')  }}" }
  - { name: daily_pf_droplasso_v6_filename, value: "{{ pf_drop_lasso_v6_filename | default('/etc/pf.drop_v6.txt')  }}" }
```

vbotka commented on tasks/drop.lasso.yml:
```diff
@@ -0,0 +1,55 @@
+    backup: no
+  when: pf_drop_lasso_enable | bool
+  
+- name: "drop.lasso: Enable periodic script"
+  community.general.sysrc:
+    name:  daily_pf_droplasso_enable
+    value: "{{ pf_drop_lasso_enable | bool | ternary('YES', 'NO') }}"
+    path:  "{{ pf_drop_lasso_periodic_file }}"
+
+- name: "drop.lasso: Enable v4"
+  community.general.sysrc:
+    name:  daily_pf_droplasso_v4
+    value: "{{ pf_drop_lasso_v4 | bool | ternary('YES', 'NO') }}"
+    path:  "{{ pf_drop_lasso_periodic_file }}"
+  when: pf_drop_lasso_enable | bool
+
+- name: "drop.lasso: Enable v6"
+  community.general.sysrc:
+    name:  daily_pf_droplasso_v6
+    value: "{{ pf_drop_lasso_v6 | bool | ternary('YES', 'NO') }}"
+    path:  "{{ pf_drop_lasso_periodic_file }}"
+  when: pf_drop_lasso_enable | bool
+
+- name: "drop.lasso: v4 file"
+  community.general.sysrc:
+    name: daily_pf_droplasso_v4_filename
+    # this variable is also set in templates/default-drop-lasso.j2
+    value: "{{ pf_drop_lasso_drop_v4_filename | default('/etc/pf.drop_v4.txt') }}"
+    path:  "{{ pf_drop_lasso_periodic_file }}"
+  when: pf_drop_lasso_drop_v4_filename is defined
+
+- name: "drop.lasso: v6 file"
+  community.general.sysrc:
+    name: daily_pf_droplasso_v6_filename
+    # this variable is also set in templates/default-drop-lasso.j2
+    value: "{{ pf_drop_lasso_drop_v6_filename | default('/etc/pf.drop_v6.txt') }}"
+    path:  "{{ pf_drop_lasso_periodic_file }}"
+  when: pf_drop_lasso_drop_v6_filename is defined
```
Apply *pf_drop_lasso_rcconf* in a loop. I think the conditions are not necessary.

```suggestion
- name: "drop.lasso: Configure periodic."
  community.general.sysrc:
    path: "{{ pf_drop_lasso_periodic_file }}"
    name: "{{ item.name }}"
    value: "{{ item.value }}"
  loop: "{{ pf_drop_lasso_rcconf }}"
```

dlangille committed:
Update defaults/main/drop_lasso.yml

Co-authored-by: Vladimir Botka <vbotka@gmail.com>

dlangille committed:
Update tasks/drop.lasso.yml

Co-authored-by: Vladimir Botka <vbotka@gmail.com>

dlangille commented:
I'll do some testing with this

dlangille commented on tasks/drop.lasso.yml:
```diff
@@ -0,0 +1,26 @@
+    path: "{{ pf_drop_lasso_periodic_file }}"
+    name: "{{ item.name }}"
+    value: "{{ item.value }}"
+  loop: "{{ pf_drop_lasso_rcconf }}"
```
This solution always sets the `filename` values. I wanted to avoid setting them unless the user specified them.

```
[13:54 gw01 dvl /etc] % cat /etc/periodic.conf.local   
daily_pf_droplasso_enable="True"
daily_pf_droplasso_v4="YES"
daily_pf_droplasso_v6="YES"
daily_pf_droplasso_v4_filename="/etc/pf.drop_v4.txt"
daily_pf_droplasso_v6_filename="/etc/pf.drop_v6.txt"
```
This is only a nice thing. Not a requirement.

vbotka replied:
I see. You can remove the *.filename defaults from the template. But, *daily_pf_droplasso_enable* should be fixed to YES/NO.

dlangille committed:
I want to reduce the number of places where filenames were hardcoded (e.g. /etc/pf.drop_v4.txt)
I wanted one place, I've settled on 2 - this is Ansible, not a coding language.

I kept getting:

TASK [pf : drop.lasso: Install periodic script] ************************************************************************************************************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: ansible.errors.AnsibleUndefinedVariable: 'daily_pf_droplasso_v4_filename' is undefined. 'daily_pf_droplasso_v4_filename' is undefined
fatal: [gw01.int.unixathome.org]: FAILED! => {"changed": false, "msg": "AnsibleUndefinedVariable: 'daily_pf_droplasso_v4_filename' is undefined. 'daily_pf_droplasso_v4_filename' is undefined"}

So I moved the definition of daily_pf_droplasso_v4_filename & daily_pf_droplasso_v6_filename
onto separate lines.

I'd prefer not to specify daily_pf_droplasso_v4_filename etc in periodic.conf
when they are default values; I gave up.

set daily_pf_droplasso_enable to YES/NO, not true/false.

vbotka commented:
LGTM. Do you also think this can be merged?

dlangille commented:
There is one change I want to make: the upstream URL from which the bogons are pulled. I have a better source.

I can't code that right now: I want to install this gateway host into the rack and get it started.

vbotka commented:
> There is one change I want to make: the upstream URL

Wouldn't you prefer to do it in a new PR? This one seems to be complicated enough :) It's devel anyway.

dlangille committed:
Change the bogon source from https://www.spamhaus.org to https://team-cymru.org

The former is a list of susspect IP ranges.
The latter is a list of IP unallocated addresses - they should not be see in routing.

I feel this more closely matches the purpose of not routing bogons.

This change means we don't need the jq dependency - it has been removed.

The IPv6 version of this file is bigger (about 150195 vs 78 lines).
As such, we need to adjust a sysctl to allow pf to handle that data.

As with period,conf, give the user the option to install that sysctl
to /etc/sysctl.conf.local

dlangille commented:
> > There is one change I want to make: the upstream URL
> 
> Wouldn't you prefer to do it in a new PR? This one seems to be complicated enough :) It's devel anyway.

Thank you, I just completed it. It removes a dependency.

vbotka commented:
Thank you for your contribution!

JFYI. I'll proceed with a few changes:

* make the DROPURL configurable (default pf_drop_lasso_dropurl: https://team-cymru.org/Services/Bogons/fullbogons-ipv)

* rename `tasks/drop.lasso.yml` to `drop-lasso.yml`

* change `name: "drop.lasso ...` to `name: "Drop.lasso ..`. Otherwise ansible-lint complains [name casing](https://ansible.readthedocs.io/projects/lint/rules/name/?h=upper#name)

* make debug tasks conditional

* update tasks/debug.yml

* update changelog.

All changes will be backward compatible. I'll come back to you before the release.

vbotka commented:
The devel branch is ready. If you don't object I'll release it on Monday.

dlangille commented:
I had considered allowing for both spamhaus and cymru - but that started getting complicated.

If the user was to change the value of `pf_drop_lasso_dropurl`, they'd probably have to modify these lines, to get the downloaded filename correct for the new download location:

* https://github.com/vbotka/ansible-freebsd-pf/blob/devel/templates/default-drop-lasso.j2#L31
* https://github.com/vbotka/ansible-freebsd-pf/blob/devel/templates/default-drop-lasso.j2#L43

But if they're doing that, they are probably savvy enough to know how to create a new template.

dlangille commented:
FYI, all this is going towards my new gateway, which I racked earlier today: https://dan.langille.org/2025/01/28/gw01/

Some some final touches to apply, but tomorrow I'll swap the cables to make it live.

vbotka commented:
> ... racked earlier today: https://dan.langille.org/2025/01/28/gw01/

Nice! As a hint: You can upload dmesg to the searchable database at https://dmesgd.nycbug.org/index.cgi

vbotka commented:
> But if they're doing that, they are probably savvy enough to know how to create a new template.

Right. Let's keep it "tools not policy".

dlangille commented:
> > ... racked earlier today: https://dan.langille.org/2025/01/28/gw01/
> 
> Nice! As a hint: You can upload dmesg to the searchable database at https://dmesgd.nycbug.org/index.cgi

I'd forgotten about that project. Done.
